lean4-infoview-api/src/rpcSessions.ts

```diff
@@ -0,0 +1,171 @@
+     * keepalive notifications.
+     */
+    createRpcSession(uri: DocumentUri): Promise<string>;
+    /** Closes an RPC session created with `createRpcSession`. */
+    closeRpcSession(sessionId: string): void;
+    /** Sends an RPC call to the Lean server. */
+    call(request: RpcCallParams): Promise<any>;
+    /** Sends an RPC reference release notification to the server. */
+    release(request: RpcReleaseParams): void;
+}
+
+/**
+ * An RPC session at a specific point in a file.
+ * The Lean 4 RPC protocol requires every request to specify a position in the
+ * file; only `@[serverRpcMethod]` declarations above this positions are callable.
+ * Implementations of this interface bundle the position.
+ */
+export interface RpcSession {
+    call(method: string, params: any): Promise<any>;
+}
+
+class RpcSessionForFile {
+    sessionId: Promise<string>;
+    /**
+     * `failed` stores a fatal exception indicating that the RPC session can no longer be used.
+     * For example: the worker crashed, etc.
+     */
+    failed?: any;
+
+    refsToRelease: RpcPtr<any>[] = [];
+    finalizers: FinalizationRegistry<RpcPtr<any>>;
+
+    constructor(public uri: DocumentUri, public sessions: RpcSessions) {
+        this.sessionId = (async () => {
+            try {
+                return await sessions.iface.createRpcSession(uri);
+            } catch (ex) {
+                this.failWithoutClosing(ex);
+                throw ex;
+            }
+        })();
+        this.sessionId.catch(() => {}); // silence uncaught exception warning
+
+        // Here we hook into the JS GC and send release-reference notifications
+        // whenever the GC finalizes a number of `RpcPtr`s. Requires ES2021.
+        let releaseTimeout: number | undefined
+        this.finalizers = new FinalizationRegistry(ptr => {
+            if (this.failed) return;
+            this.refsToRelease.push(ptr)
+
+            // We release eagerly instead of delaying when this many refs become garbage
+            const maxBatchSize = 100
+            if (this.refsToRelease.length > maxBatchSize) {
+                void this.releaseNow();
+                clearTimeout(releaseTimeout);
+                releaseTimeout = undefined;
+            } else if (releaseTimeout === undefined) {
+                releaseTimeout = setTimeout(() => {
+                    void this.releaseNow()
+                    releaseTimeout = undefined;
+                }, 100);
+            }
+        });
+    }
+
+    async releaseNow() {
+        const sessionId = await this.sessionId;
+        if (this.failed || this.refsToRelease.length === 0) return;
+        this.sessions.iface.release({
+            uri: this.uri,
+            sessionId,
+            refs: this.refsToRelease,
+        });
+        this.refsToRelease = [];
+    }
+
+    register(o: any) {
+        if (o instanceof Object) {
+            if (Object.keys(o as {}).length === 1 && 'p' in o && typeof(o.p) !== 'object') {
```

Comment:
I think we have very different opinions here.  I find it hard to see the added value of having type-specific registration code when a ten-line function does the same job as well.  With the previous design we also had to expose the `registerRefs` function, adding to the API surface.

There is one more advantage of doing the registration automatically: it works nicely with dynamic code where we don't know the type, such as with the widget props.  With the automatic registration, the props could contain RPC references and the `getWidgets` call will automatically register them.

Reply:
For me the perceived advantage is type-safety. With this variant, I actually can't send over an object with the same shape (i.e. `{p: "..."}`), or I'll get a bunch of weird "unknown reference" errors. Is that likely to happen in practice? Probably not, so I think this solution is fine, but with more typed approaches we don't have to worry about this kind of issue.

Reply:
> With the automatic registration, the props could contain RPC references and the `getWidgets` call will automatically register them.

They actually still can't since widget instance props are stored in the `InfoTree` which cannot meaningfully contain `RpcRef`s, as there is no associated RPC session (there might not even be a server).

Reply:
I guess this is the old debate of conservative vs. exact garbage collection, except now with JSON. :smile: 

> They actually still can't since widget instance props are stored in the InfoTree which cannot meaningfully contain RpcRefs, as there is no associated RPC session (there might not even be a server).

This is not a big issue.  We can just store a thunk there.  This means you'll get different references every time you call `getWidgets`, but we've got the same situation with `getInteractive{Messages,Goals,TermGoal}` already.

Reply:
> Is that likely to happen in practice?

Not if we exclusively use serialized Lean types (because in the worst case we can just make the derive handler complain about the field name).  But arbitrary `Json` fields are a bit of an issue:
```lean
structure Foo where
  randomJsonData : Json
  deriving FromJson, ToJson
```

Reply:
> This is not a big issue. We can just store a thunk there.

Interesting idea. The current storage type for widget instances is `Json`, so I guess we would store something like `[MonadRpcSession m] Unit -> m Json`?

Reply:
Right, I was rather thinking about `StateM RpcObjectStore Json` (with the types from https://github.com/leanprover/lean4/pull/1356) so that it lives in `Type 0`.